ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -271,4 +489,376 @@ private void updateMapHelper(Map<String, Long> map, String key, Long value) {
+          Map.Entry<Offset, IndexSegment> activeIndexSegmentEntry = indexSegments.floorEntry(firstCheckpoint);
+          long indexSegmentStartProcessTime = time.milliseconds();
+          List<IndexEntry> activeIndexEntries =
+              getFilteredIndexEntries(activeIndexSegmentEntry.getValue(), firstCheckpoint);
+          processIndexSegmentEntriesBackward(activeIndexSegmentEntry.getValue(), activeIndexEntries, deletedKeys);
+          metrics.statsBucketingScanTimePerIndexSegmentMs.update(time.milliseconds() - indexSegmentStartProcessTime,
+              TimeUnit.MILLISECONDS);
+          if (indexSegments.size() > 1) {
+            ConcurrentNavigableMap<Offset, IndexSegment> sealedIndexSegments =
+                indexSegments.subMap(index.getStartOffset(), activeIndexSegmentEntry.getKey());
+            for (IndexSegment indexSegment : sealedIndexSegments.descendingMap().values()) {
+              if (cancelled) {
+                return;
+              }
+              indexSegmentStartProcessTime = time.milliseconds();
+              List<IndexEntry> indexEntries = getIndexEntries(indexSegment);
+              processIndexSegmentEntriesBackward(indexSegment, indexEntries, deletedKeys);
+              metrics.statsBucketingScanTimePerIndexSegmentMs.update(time.milliseconds() - indexSegmentStartProcessTime,
+                  TimeUnit.MILLISECONDS);
+            }
+          }
+        } else {
+          newScanResults.updateLogSegmentBaseBucket(index.getStartOffset().getName(), 0L);
+        }
+        recentEntryQueueEnabled = true;
+        Offset secondCheckpoint = index.getCurrentEndOffset();
+        if (secondCheckpoint.compareTo(firstCheckpoint) > 0) {
+          forwardScan(firstCheckpoint, secondCheckpoint);
+        }
+        newScanResults.scannedEndOffset = secondCheckpoint;
+        scanResults.set(newScanResults);
+      } catch (Exception e) {
+        logger.error("Exception while scanning index for stats bucketing", e);
+        metrics.blobStoreStatsErrorCount.inc();
+      } finally {
+        scanLock.lock();
+        isScanning = false;
+        waitCondition.signalAll();
+        scanLock.unlock();
+        metrics.statsBucketingScanTotalTimeMs.update(time.milliseconds() - startTimeInMs, TimeUnit.MILLISECONDS);
+      }
+    }
+
+    /**
+     * Perform a forward scan (order to newest entries) from the given start offset to the given end offset. This
+     * forward scan is used to process any new entries that were added to the index while scanning to the first
+     * checkpoint.
+     * @param startOffset the start offset defining the start of the scan range, inclusive (greater than or equal to)
+     * @param endOffset the end offset defining the end of the scan range, exclusive (strictly less than)
+     * @throws StoreException
+     */
+    private void forwardScan(Offset startOffset, Offset endOffset) throws StoreException {
+      SortedMap<Offset, IndexSegment> tailIndexSegments =
```

Comment:
can we have a metric to track the range of forward scan. It might give us an idea on how to tweak the queueProcessingTime wrt memory consumption.